Improve the contributing guide (#1045)
fingerprintjs/fingerprintjs: contributing.md

# Contributing to FingerprintJS

Thanks for taking the time to contribute!
Here you can find ways to make FingerprintJS better, as well as tips and guidelines.

This project and everyone participating in it is governed by the [Code of Conduct](code_of_conduct.md).
By participating, you are expected to uphold this code.

## How you can contribute

### Reporting an issue

If you've noticed a bug, have an idea or a question,
feel free to [create an issue](https://github.com/fingerprintjs/fingerprintjs/issues/new/choose) or [start a discussion](https://github.com/fingerprintjs/fingerprintjs/discussions/new/choose).

Before you start, please [search](https://github.com/search?q=repo%3Afingerprintjs%2Ffingerprintjs&type=code) for your topic.
There is a chance it has already been discussed.

When you create an issue, the description is pre-filled with a template text.

> Please fill the missing information carefully, it will help us solve your issue faster.

Reviewer: ```suggestion
Please fill in the missing information carefully, it will help us solve your issue faster.
```

If you want share a piece of code or the library output with us, please wrap it in a ` ``` ` block and make sure you include all the information.

### Creating a pull request

If you want to fix a bug, add a source of entropy or make any other code contribution, please [create a pull request](https://docs.github.com/en/get-started/exploring-projects-on-github/contributing-to-a-project).

> After you clone the repository, check the [Working with code](#working-with-code) section to learn how to run, check and build the code.

Reviewer: There are still a few missing oxford commas and grammar typos in the document. Consider running the text through Grammarly to catch them all at once :) 

```suggestion
After you clone the repository, check the [Working with code](#working-with-code) section to learn how to run, check, and build the code.
```

Author: Feel free to add all the missing commas by pushing a commit

Reviewer: [Done](https://github.com/fingerprintjs/fingerprintjs/pull/1045/commits/3ec1fc2b361a8455f8c48a6a1cb122dfaf6602c5) and approved, thank you!

Author: Thank you for the improvements!

In order for us to review and accept your code contributions, please follow these rules:

> - Your code quality should be at least as good as the code that you modify;

Reviewer: Some list items in the document end with `.` and some end with `;`. Periods are [more](https://developers.google.com/style/lists) [conventional](https://learn.microsoft.com/en-us/style-guide/scannable-content/lists), I would unify it and end all list item sentences with a period.

```suggestion
- Your code quality should be at least as good as the code you modify.
```

- Your code style (syntax, naming, coding patterns, etc) should follow the FingerprintJS style.
- All the new code should be covered with automated tests.
- All the checks described in the [Working with code](#working-with-code) section must pass successfully.
    You may create a draft pull request in this repository to run the checks automatically by GitHub Actions,
    but the tests won't run on BrowserStack until a FingerprintJS maintainer approves them.
- If you want to add an entropy source (component), follow the [How to add an entropy source](#how-to-add-an-entropy-source) instructions carefully.

> - The changes should be backwards compatible, ensuring FingerprintJS users continue to use the library without any modifications.

Reviewer: ```suggestion
- The changes should be backward compatible, ensuring FingerprintJS users continue to use the library without any modifications.
```

- Don't add dependencies (such as Node packages) unless necessary;

> - Don't make changes unrelated the stated pull request purpose. Please strive to introduce as few changes as possible;

Reviewer: ```suggestion
- Don't make changes unrelated to the stated purpose of your pull request. Please strive to introduce as few changes as possible;
```

- Don't change FingerprintJS code style, its TypeScript configuration or other subjective things.

If you want to do something more complex than fixing a small bug, or if you're not sure if your changes meet the project requirements, please [start a discussion](https://github.com/fingerprintjs/fingerprintjs/discussions/new/choose).
We encourage starting a discussion if you want to propose changing a rule from this guide.
Doing so ensures we discuss all opinions, creating a good contribution experience for everyone.

### Helping with existing issues

If you want to help, but don't know where to start, take a look at the ["help wanted" issues](https://github.com/fingerprintjs/fingerprintjs/labels/help%20wanted).
You can help by sharing knowledge or creating a pull request.
Feel free to ask questions in the issues if you need more details.

## Working with code

> This section describes how to deploy the repository locally, make changes to the code and verify your work.

Reviewer: ```suggestion
This section describes how to deploy the repository locally, make changes to the code, and verify your work.
```

First, make sure you have [Git](https://git-scm.com), [Node.js](https://nodejs.org/en) and [Yarn](https://yarnpkg.com) installed.
Then clone the repository and install the dependencies:

```bash
git clone https://github.com/fingerprintjs/fingerprintjs.git
cd fingerprintjs
yarn install
```

### Development playground

> Development playground allows to run FingerprintJS locally. Run this command to start a playground:

Reviewer: ```suggestion
Development playground lets you run FingerprintJS locally. Run this command to start a playground:
```
```suggestion
Development playground lets you run FingerprintJS locally. Run this command to start a playground:
```

Before you start, please [search](https://github.com/search?q=repo%3Afingerprintjs%2Ffingerprintjs&type=code) for your topic.
There is a chance it has already been discussed.

When you create an issue, the description is pre-filled with a template text.
Please fill the missing information carefully, it will help us solve your issue faster.
If you want to show a code or a FingerprintJS output, wrap it in ` ``` ` blocks and don't truncate.

### Creating a pull request

If you want to fix a bug, add an entropy source or make another change to the code, you can [create a pull request](https://docs.github.com/en/get-started/exploring-projects-on-github/contributing-to-a-project).

After you clone the repository, check the [Working with the code](#working-with-the-code) section to learn how to run, check and build the code.

Please follow these guidelines so that we can accept your code changes:
- Your code quality should be at least as good as the code that you modify;
- Your code style (syntax, naming, coding patterns, etc) should follow the style of the other repository code;

> - The new code should be covered with automatic tests;

Reviewer 2: All the new code should be covered with automated tests

- All the checks described in the [Working with the code](#working-with-the-code) section must pass successfully.
    You may create a draft pull request in this repository to run the checks automatically by GitHub Actions,

>     but the tests won't run on BrowserStack until a FingerprintJS maintainer approves it;

Reviewer 2: approves them or approves the pull request.

- If you want to add an entropy source (fingerprint component), follow the [How to make an entropy source](#how-to-make-an-entropy-source) instructions carefully;

> - The change shouldn't be breaking, that is FingerprintJS users don't need to change code on their side after your changes;

Reviewer 2: The changes should be backwards compatible, ensuring FingerprintJS users continue to use the library without any modifications.

- Don't add dependencies (such as Node packages) unless necessary;

> - Don't make changes not related to the PR goal, try to keep your changes as small as possible;

Reviewer 2: ..unrelated the stated pull request purpose. Please strive to introduce as few changes as possible.

> - Don't change opinionated things, such as the code style or TypeScript configuration.

Reviewer 2: I'd simply here as "Don't change FingerprintJS code style or its TypeScript configuration"

Author: I'd like to keep a generalized term in this sentence. I've changed to "Don't change FingerprintJS code style, its TypeScript configuration or other subjective things".

> If you want to do something more complex than fixing a small bug, or unsure whether you changes will meet the requirements, or want to violate a guideline,

Reviewer 2: or if you're not sure if your changes meet the project requirements, please start a discussion. 
We encourage starting a discussion if you want to propose changing a rule from this guide. 
Doing so ensures we discuss all opinions, creating a good contribution experience for everyone."

please [start a discussion](https://github.com/fingerprintjs/fingerprintjs/discussions/new/choose) before making a pull request.
It will let you avoid wasting your time, and will improve your contributing experience.

### Helping with existing issues

> If you want to contribute but don't know what to improve,

Reviewer 2: If you want to help, but don't know where to start, take a look at the "help wanted" issues[link to open issues].

check [the open issues](https://github.com/fingerprintjs/fingerprintjs/issues),
especially those that are marked with a [help wanted](https://github.com/fingerprintjs/fingerprintjs/labels/help%20wanted) label.
You can help by sharing knowledge or creating a pull request.
Feel free to ask questions in the issues if you need more details.

> ## Working with the code

Reviewer 2: Should be 'Working with code' or 'Working with the repository code'

> This section describes how to deploy the repository locally, change the code and check your changes.

Reviewer 2: .. make changes to the code and verify your work.

First, make sure you have [Git](https://git-scm.com), [Node.js](https://nodejs.org/en) and [Yarn](https://yarnpkg.com) installed.

> Clone the code and install the dependencies:

Reviewer 2: Then clone the repository and install the dependencies.

```bash
git clone https://github.com/fingerprintjs/fingerprintjs.git
cd fingerprintjs
yarn install
```

### Development playground

Development playground allows to run FingerprintJS locally. Run this command to start a playground:

```bash
yarn playground:start # Add '--port 8765' to change the server port
```

Then open http://localhost:8080 in a browser.
FingerprintJS will execute immediately and print the result on the page.
The page reloads every time you change the source code.
The code of the playground itself is located in the [playground](playground) directory.

If you want to open the playground on a device outside your local network, use a tool like [Ngrok](https://ngrok.com).

> To build the playground distribution code (e.g. to upload to a static server), run:

Reviewer 2: e.g. to upload it to a web server

### How to build

> To build the distribution files of FingerprintJS, that can be used in browser directly, run:

Reviewer: ```suggestion
To build the distribution files of FingerprintJS that can be used in the browser directly, run:
```

yarn check:ssr
```

### How to add an entropy source

An entropy source is a function that gets a piece of data about the browser.

> The value returned by an entropy source is used to produce the visitor identifier and called an entropy component.

Reviewer: ```suggestion
The value returned by an entropy source (also called an entropy component) is used to produce the visitor identifier.
```

Entropy sources are located in the [src/sources](src/sources) directory.
All entropy components must be simple JavaScript values that can be encoded into JSON.

Entropy source must meet the following requirements:

> - It is stable, that is always or almost always produces the same value in each browser, including incognito, guest and desktop modes;

Reviewer: I am not sure this "that is" phrasing is grammatically correct, at least I bumped on it. Also missing oxford comma. 

```suggestion
- It is stable — it always or almost always produces the same value in each browser, including incognito, guest, and desktop modes;
```

Author: Oxford comma is optional (hence the name), isn't it?

Author: Looks like it inclines to be required in American English, and we use it in other documents, so let's add it

> - It is selective, that is produces different values in different browsers, OSs or devices.

Reviewer: ```suggestion
- It is selective — it produces different values in different browsers, operating systems, or devices.
```

    A good entropy source represents the browser, operating system, or device settings.
- It produces no side effects, such as messages in the browser console, DOM changes, modal windows, notifications, and sounds.
- It is fast. An entropy source should take not more than 1 second to complete.
- It doesn't represent only the browser version. An example of such a signal is a JavaScript feature probing.
    It's not a good entropy source because the return value is likely to change due to automatic browser updates,
    the selectivity is low, and the same entropy can be achieved by just using the User-Agent string.

> Entropy source runs in 2 stages: "load" and "get":

Reviewer: ```suggestion
The entropy source runs in 2 stages: "load" and "get":
```

Author: I suppose it's wrong because the text refers to entropy source as a concept, not a specific entropy source

> Entropy sources are located in the [src/sources](src/sources) directory.

Reviewer 2: Cannot make the comment on line 192, but I'd say "How to add an entropy source"

> Entropy component must be a simple JS value that can be JSON-encoded.

Reviewer 2: I think we should be consistent with calling it wither "Entropy sources" or "entropy components" and avoid using multiple alternatives, e.g. "Fingerprint component". 
Also: "Entropy sources must be simple JavaScript values that can be encoded into JSON".

Author: IIRC you used terms "entropy source" (a function) and "fingerprint component" (the return value) in v2. We keep this terminology, but we've changed "fingerprint component" to "entropy component" because we use `visitorId` instead of `fingerprint` in v3 to keep API compatibility with the Pro version. So I'd like to keep both "entropy source" and "entropy component" and remove "fingerprint component".

I don't mind using a single term for both function and its return value (we use "signal" internally), but I think it should be a separate PR because it's an extensive change.

> Entropy source must meet the following requirements:

Reviewer 2: An entropy source is a ..
Entropy sources must meet..

We should use a singular with an "An" article or a plural with no article, but not a mix between the two.

async function entropySource() {
  const finalData = await doLongAction()

>   // If the source's returned value is not a function, it's considered as an entropy component

Reviewer: ```suggestion
  // If the source's returned value is not a function, it's considered an entropy component
```

```

> Entropy source must handle expected and only expected errors.

Reviewer: ```suggestion
Entropy sources must handle expected and only expected errors.
```
